### DISCUSSION TITLE
How can I load a static HTML file from the public directory into my index.js file in the pages directory?

### DISCUSSION THREAD

**Discussion by OwaisSiddiqui** [vercel/next.js#12373] [Help] [answered]
I am new to NextJS and ReactJS. I am trying to deploy a listings website (such as craigslist.org or zillow.com). I have a landing page as a static pure HTML page with some CSS/JS. However, I am not sure how to load this from my `index.js` file. I have the following currently:

`index.js`:
`import "../public/index.html"
export default function IndexHTML() {
    return ??? <-- Not sure what to do here
}`

and I get the following error (considering I replace `??? <-- Not sure what to do here` with `<div></div>`:

`./public/index.html 1:0
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> <!DOCTYPE html>
| <html>
| <head>`

Someone from the nextjs Reddit suggests I paste the HTML as a react and return that. But I pretty sure you can serve static files as HTML (from this https://nextjs.org/docs/basic-features/static-file-serving, though I am not sure how to do it for HTML files).

**Comment by leerob** [comment]
You'll likely want to use [JSX to render HTML through React](https://reactjs.org/docs/introducing-jsx.html). This will allow you to add interactively to your HTML through React.

**Comment by alexfinnarn** [comment, marked as answer]
I have a need to serve some static HTML created in another project but mostly still use Next.js as usual. With static file serving, you can put whatever HTML file you want in `/public` and reach it, ex. `/public/foo.html` -> `https://my.site/foo.html`.

Okay, you say, but I don't want to make users use `.html` as without that extension you get a 404. Well, you can use redirect rules in `next.config.js`. For a more complicated setup, you can use an Express server in front Next.js...I think that is in the official docs.

I wouldn't do anything with JSX/React though and stick to Next.js' static file serving + redirects or a more involved front controller in Express/Koa/whatevs to read and serve static HTML files before hitting Next.js.

**Reply by esemeniuc** [reply]
You can add rewrites like this to your `next.config.js`. To allow the url `mysite.com/dir/customPage` to point to `/public/dir/customPage/index.html`, I added this to my config (with the content in my `/public` directory):

```
rewrites: async () => {
    return [
      {
        source: "/dir/:customPage",
        destination: "/dir/:customPage/index.html",
      }
    ]
}
```

**Reply by pondorasti** [reply]
This answer just saved me a lot for time! I noticed that you can use relative paths to load images (`<img src="cat.png"`), but is there a way to load external stylesheets? The default way of adding `<link rel="stylesheet" href="./style.css">` to `<head>` doesn't work for me. For now, I've gotten around this issue by inlining my css....

**Reply by esemeniuc** [reply]
You can also load from `<link rel="stylesheet" href="/style.css">`, and that would load from `myApp/public/style.css`

**Reply by emirhansirkeci** [reply]
> You can add rewrites like this to your `next.config.js`. To allow the url `mysite.com/dir/customPage` to point to `/public/dir/customPage/index.html`, I added this to my config (with the content in my `/public` directory):
> 
> ```
> rewrites: async () => {
>     return [
>       {
>         source: "/dir/:customPage",
>         destination: "/dir/:customPage/index.html",
>       }
>     ]
> }
> ```

I want to use my index.html as my homepage. When user hits this link "https://mysite.com/" i want to render my html. How can i do that too with this config?

**Reply by bharathmsd7** [reply]
Is there any way to use the css and js from public folder in HTML 
project structure =>
public 
 - index.html
 - css
    - styles.css
 - js
    - script.js

Need to access those CSS and JS files inside index.html, tried normal ways doesn't help me.

**Reply by deadcoder0904** [reply]
@justChargin read https://blog.bitsrc.io/html-pages-in-nextjs-using-rewrites-bc5f56ea3ed

you can do it with:

```ts
rewrites: async () => {
    return [
      {
        source: '/',
        destination: '/index.html',
      },
    ]
},
```

make sure you don't have `pages/index.tsx` as it will conflict with `/`... again read the blog... it explains better.

@bharathmsd7 just put `link` & `script` tag inside the `index.html`... doesn't have to be complicated.

**Reply by bharathmsd7** [reply]
Thanks, @deadcoder0904 it worked for me!

**Reply by alireza1691** [reply]
> This answer just saved me a lot for time! I noticed that you can use relative paths to load images (`<img src="cat.png"`), but is there a way to load external stylesheets? The default way of adding `<link rel="stylesheet" href="./style.css">` to `<head>` doesn't work for me. For now, I've gotten around this issue by inlining my css....

Where did you put it?

**Reply by krthush** [reply]
> @justChargin read https://blog.bitsrc.io/html-pages-in-nextjs-using-rewrites-bc5f56ea3ed
> 
> you can do it with:
> 
> ```ts
> rewrites: async () => {
>     return [
>       {
>         source: '/',
>         destination: '/index.html',
>       },
>     ]
> },
> ```
> 
> make sure you don't have `pages/index.tsx` as it will conflict with `/`... again read the blog... it explains better.
> 
> @bharathmsd7 just put `link` & `script` tag inside the `index.html`... doesn't have to be complicated.

This was a life saver!! Thanks @deadcoder0904 

By using this you are able to easily host landing pages from other sources such as Webflow (code export) and then use the Next.JS server/logic for the rest of your application 🚀

**Reply by mohasinkr** [reply]
I tried the same things, setup app rewrites, placed the static files inside public folder etc. The routing works fine on localhost and while running npm build on local system. When I try to host the application on AWS amplify getting 404 error for that route. Rest of the application loads fine. Any possible solutions for this?

**Reply by jjrh92** [reply]
@deadcoder0904 Good day sir, what if i want to serve those static files from a source folder other than / (app/page) let's say i create a static folder outside of pages at the root level of app? Thank you in advance.

**Comment by chrisandrew** [comment]
The following should work: place the static HTML file in the public folder and use the <a../> anchor element in your your pages and/or components to navigate to it.

**Comment by votemike** [comment]
I would like my Storybook available on the same domain as my Next App, what would be the best way to do that? Get Storybook to build out to /public/styleguide?
Or is there a better way to serve non-NextJs pages through the Next routing?

**Reply by drmanojgrg** [reply]
none of the answers worked for me.

**Reply by votemike** [reply]
This is what I ended up doing: https://github.com/votemike/test-next-js/commit/d55aca5c65fa2b8b037d3a13dcd3930c4a1b724f

**Comment by yahya-aghdam** [comment]
Super easy way is pasting your html code with css and js in `public` folder, so you can route to `mywebsite.com/index.html` with no longer code :)

You can handle main url redirection from `next.config` file with adding this:
YOU HAVE TO DELETE MAIN INDEX.TSX OR PAGE.TSX FROM SRC OR APP FOLDER
```JavaScript
rewrites: async () => [
    {
      source: "/",
      destination: "/index.html",
    },
  ],
```

**Comment by startupsandcode** [comment]
If you want to have multiple pages and not write a rewrite for every specific file:

In your public folder create an html folder and you can do this to have any html file be there

```javascript
async rewrites() {
  return [
	{
		source: "/html/:slug",
		destination: "/html/:slug.html",
	}
  ]
},
```
